libdrgn/lazy_parameter.c

```diff
@@ -5,31 +5,60 @@
 
 void drgn_lazy_parameter_deinit(struct drgn_lazy_parameter *lazy_parameter)
 {
-	if (!drgn_lazy_parameter_is_evaluated(lazy_parameter))
-		drgn_type_thunk_free(lazy_parameter->thunk);
+	if (lazy_parameter->is_object) {
+		if (!drgn_lazy_parameter_is_evaluated(lazy_parameter))
+			drgn_object_thunk_free(lazy_parameter->object_thunk);
+		else
+			free(lazy_parameter->object);
+	} else
+		if (!drgn_lazy_parameter_is_evaluated(lazy_parameter))
+			drgn_type_thunk_free(lazy_parameter->type_thunk);
 }
 
 inline struct drgn_error *
 drgn_lazy_parameter_check_prog(struct drgn_lazy_parameter *lazy_parameter,
 			       struct drgn_program *prog)
 {
-	if ((drgn_lazy_parameter_is_evaluated(lazy_parameter) ?
-	     drgn_type_program(lazy_parameter->type) :
-	     lazy_parameter->thunk->prog) != prog) {
-		return drgn_error_create(DRGN_ERROR_INVALID_ARGUMENT,
-					 "type is from different program");
+	if (lazy_parameter->is_object) {
+		if ((drgn_lazy_parameter_is_evaluated(lazy_parameter) ?
+		     drgn_type_program(lazy_parameter->object->type) :
+		     lazy_parameter->object_thunk->prog) != prog) {
+			return drgn_error_create(DRGN_ERROR_INVALID_ARGUMENT,
+						 "object is from different program");
+		}
+	} else {
+		if ((drgn_lazy_parameter_is_evaluated(lazy_parameter) ?
+		     drgn_type_program(lazy_parameter->type) :
+		     lazy_parameter->type_thunk->prog) != prog) {
+			return drgn_error_create(DRGN_ERROR_INVALID_ARGUMENT,
+						 "type is from different program");
+		}
 	}
 	return NULL;
 }
 
+struct drgn_error
+*drgn_lazy_parameter_evaluate_type(struct drgn_lazy_parameter *parameter,
+				   struct drgn_qualified_type *ret) {
```

Comment:
It's confusing having two functions that get the type of a lazy parameter. Can we make `drgn_lazy_parameter_get_type()` handle both objects and types?